linera-core/src/chain_worker/actor.rs

```diff
@@ -307,6 +340,61 @@ where
             }
         }
 
+        self.service_runtime_thread
+            .await
+            .expect("Service runtime thread should not panic");
         trace!("`ChainWorkerActor` finished");
     }
+
+    /// Clear the current [`QueryContext`] if any changes might be made, which will force a restart
+    /// of the service runtime actor the next time before a query is handled.
+    fn maybe_clear_query_context(&mut self, request: &ChainWorkerRequest<StorageClient::Context>) {
```

Comment:
Only a (successful) call to `ProcessConfirmedBlock` can modify the execution states of the applications.

Reply:
I agree, except for `HandleBlockProposal`: That one also needs to simulate execution, and needs to start from a pristine state.

So before _and after_ `HandleBlockProposal`, the context needs to be cleared, if I understand correctly. (But I think the "after" part is taken care of in `prepare_to_query_application`.)

Reply:
I'm wondering if it might make more sense to move the service runtime actor into `ChainWorkerState` instead, and restart it every time `save` is called :thinking:

Reply:
But after `HandleBlockProposal` `save` isn't called.

Reply:
I'm not sure I see the issue with the block proposal. We already clone a readonly view for graphql queries

Reply:
Right, if that works with the current setup (staging the execution while also preserving the read-only view) that's even better. :+1:

Reply:
Can the chain actor own the service actor? or at least keep a handle to send a message to terminate (but there might be a race condition)

Reply:
I think so. We can start the thread in the actor, and send only the channel endpoint to `ChainWorkerState`.

Reply:
Done. As we discussed during the meeting, the plan is to make it so that the query can run concurrently to other worker operations. I'll do that in a separate PR to avoid growing this one's scope too much :+1:

Reply:
@jvff Do we have a task for the next iteration then?

Reply:
We do now :D I opened #2237 and added a TODO inside the `ChainWorkerActor`. Thanks!